Band-Aid for libretro Savestate load

m4xw opened the pull request:
Proper would be using PPSSPP's Queue.
This will need frontend extensions to do it sanely.
Fixes #11429 (mitigate for now)

hrydgard commented on libretro/libretro.cpp:
```diff
@@ -705,19 +714,32 @@ struct SaveStart {
 } // namespace SaveState
 
 size_t retro_serialize_size(void) {
+	// Not pausing the EmuThread here should be somewhat safe, for the time being
```
If retro_serialize_size and retro_serialize are called together to find out the size of memory and then serialize, you should try to get them inside the same "EmuThreadPause/EmuThreadStart" block...

m4xw replied:
I did consider just pausing it there and then remove the pause from serialize (or just leaving it since it already checks if its paused), but I wouldn't be able to continue there, because it would advance the frames.
It's the same issue in the other serialize functions, EmuThreadPause will advance by a frame, so it's already guaranteed to be out of sync.
The real question is if the Savestate Buffer ever shrinks, since only that should cause actual Issues?
Alternative is a hard crash right now, either way.
But I can look into that after work if you want.

m4xw replied:
Now that I think about it, I should probably wrap them in if(useEmuThread) too, since it's only enabled for GL it seems.
It should be required for all renderer`'s anyway, right?

m4xw committed:
Band-Aid for libretro Savestate load
Proper would be using PPSSPP's Queue.
This will need frontend extensions to do it sanely.
Fixes #11429 (mitigate for now)

unknownbrackets commented:
Hmm looking at this again, I'd suggest using Core_EnableStepping.  That would work on all backends.  Stepping is in other words debugging, and while debugging the CPU is paused regardless of thread behavior.

Example:

```c++
static serializeWasStepping = false;

size_t retro_serialize_size(void) {
	serializeWasStepping = Core_IsStepping();
	Core_EnableStepping(true);
	// If we're on the CPU thread actually, this may just stall for 16ms...
	Core_WaitInactive(16);

	// Intentionally leaving in stepping.
	// TODO: Not sure if 0x800000 alignment is necessary (that's 8MB.)
	return CChunkFileReader::MeasurePtr(state);
}

bool retro_serialize(void *data, size_t size) {
	if (Core_IsActive()) {
		ERROR_LOG(SYSTEM, "retro_serialize_size() must be called before retro_serialize()");
		return false;
	}
	if (size < CChunkFileReader::MeasurePtr(state)) {
		ERROR_LOG(SYSTEM, "retro_serialize() called with too small a size");
		return false;
	}

	bool retVal = CChunkFileReader::SavePtr(...);

	if (!serializeWasStepping) {
		Core_EnableStepping(false);
		// No need to wait for active, it'll resume on its own.
	}

	return retVal;
}
```

That said, probably doing this thread pausing stuff will work too, at least for GL.  Would have to save the std::thread id or something and compare to know if `Core_WaitInactive` needs to be called, I suppose.

-[Unknown]

hrydgard commented:
Merging as-is because the workaround apparently is highly in demand - @m4xw has promised to try and switch to EnableStepping later.

orbea commented:
@m4xw Thanks a lot for taking time to work on ppsspp issues!

NeoNight1986 commented:
I am pretty confused, what do I need to do to fix the safe state freezing?
